src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,573 @@
+
+        auto docs = corpus::make_corpus(cfg_);
+
+        while (iterations_left > 0)
+        {
+            std::uniform_int_distribution<std::size_t> next_random;
+            int64_t word_counter = 0;
+
+            corpus::parallel_consume(
+                *docs, pool,
+                [&]() {
+                    return sgns_local_buffer{*stream};
+                },
+                [&](sgns_local_buffer& buffer, const corpus::document& doc) {
+                    unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+                    sgns_net_vector neu1e(vector_size_, 0);        
+                    std::default_random_engine engine(seed);
+                    std::uniform_int_distribution<std::size_t> next_random;
+                    buffer.stream_->set_content(analyzers::get_content(doc));
+
+                    // Holds the target word and surrounding context words.
+                    sgns_window window;
+
+                    // Load target word.
+                    add_next_index(window, buffer.stream_, word_counter, engine, next_random);
+
+                    // Initialize future context words.
+                    while (window.size() < (max_window_size_ + 1))
+                        if (!add_next_index(window, buffer.stream_, word_counter, engine, next_random))
+                            break;
+
+                    // When the window is first loaded, the target word is the
+                    // first element in the window. The next time a word is added
+                    // to the window, this index will increase by one. This will
+                    // continue until this index reaches the middle of the window,
+                    // where it will stay until the algorithm completes.
+                    sgns_window::size_type target_widx = 0;
+
+                    // While there are target words left in the window...
+                    while (!window.empty() && (target_widx < window.size()))
+                    {
+                        // Sample a random window size.
+                        const auto window_size = next_random(engine) % max_window_size_;
+
+                        // Sweep across the window.
+                        for (sgns_window::size_type w = window_size; w < (max_window_size_ * 2 + 1 - window_size); ++w)
+                        {
+                            // Don't process a target word as its own context word.
+                            // Max window size works here because it is window size + 1th index
+                            // and therefore the target word
+                            if (w != max_window_size_)
+                            {
+                                const int64_t context_widx = (int64_t)target_widx - (int64_t)max_window_size_ + (int64_t)w;
+
+                                if (context_widx < 0 || context_widx >= (int64_t)window.size())
+                                    continue;
```

Comment:
(It's been a little while so bear with me.)

I recall trying better ways as you've described, running into issues (perhaps unrelated), and falling back to word2vec style which is what you see here. It's probably worth @mk3890 giving another shot at simplifying it. And it'll either work in which case hooray, or we'll relearn why we had to do this to begin with.

`window_size` is indeed a bad name for the variable, because like you pointed out, it's really leading to a window size of `max_window_size_ - window_size`.